src/ChecksumChallenge/Checksum/Pro2.cs

```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ChecksumChallenge;
+
+internal partial class Checksum
+{
+    public static uint ChecksumPro2(ReadOnlySpan<byte> arr)
+    {
+        if (arr.Length == 0) return 0;
+
+        uint sum = 0;
+        var i = 0;
+        for (i = 0; i <= arr.Length - 4; i+=4)
```

Comment:
Here's where you fixed the error I mentioned. If you check the ASM generated you'll see there's no bound checks inside the loop since the compiler safely determined the indexer will never go out of bounds.

Reply:
I dare saying that this actually brings more performance advantage than the shifting. But you can check and confirm.

Reply:
That's what I thought. I tried it on the original "Pro" and it ran in 220 us on my system - ~2/3 the time of the bounds checked one, but still double Senior's pointer version. Seems to be because the code generated is _awful_ - calculating z+4 etc as lea instead of in the mov:

       lea       edi,[r11+4]
       movsxd    rdi,edi
       movzx     edi,byte ptr [rdx+rdi]
       add       eax,edi

vs

       movzx     ebx,byte ptr [rdx+rdi+4]
       add       eax,ebx

Reply:
That's probably because z+4 is an int. Make z as **uint** and it will probably merge into movzx

Reply:
Except not, because Span's indexer takes an int!

So making z a uint means you have to cast `z+4` back to an int, which then means it can't eliminate the bounds checks again because it can't be sure if you're exceeding the bounds any more :)

Reply:
This was .net 6 though. .net 7 is supposed to be much better, should try that some time.

Reply:
Then we are out of luck indeed. 
But this 

```
if (arr.Length == 0) return 0;

        uint sum0 = 0, sum1 = 0, sum2 = 0, sum3 = 0;
        int z = 0;

        while (z < arr.Length - 16)
        {
            sum0 += (uint)(arr[z + 0] + arr[z + 4] + arr[z + 8] + arr[z + 12]);
            sum1 += (uint)(arr[z + 1] + arr[z + 5] + arr[z + 9] + arr[z + 13]);
            sum2 += (uint)(arr[z + 2] + arr[z + 6] + arr[z + 10] + arr[z + 14]);
            sum3 += (uint)(arr[z + 3] + arr[z + 7] + arr[z + 11] + arr[z + 15]);

            z += 16;
        }
```
here to me will compile to 

```
       lea      edi, [r11+01H]
       movzx    rdi, byte  ptr [rdx+rdi]
       add      r8d, edi
```
Note the lack of the extra `movsxd    rdi,edi`. I wonder why your example has the extra sign extension instruction

Reply:
```if (arr.Length == 0) return 0;

        uint sum0 = 0, sum1 = 0, sum2 = 0, sum3 = 0;
        nint z = 0;

        ref byte reference=ref MemoryMarshal.GetReference(arr);
           
        while ((int)z <= arr.Length - 4)
        {           
            sum0 += Unsafe.Add(ref reference, z);
            sum1 += Unsafe.Add(ref reference, z+ (nint)1);
            sum2 += Unsafe.Add(ref reference, z+ (nint)2);
            sum3 += Unsafe.Add(ref reference, z+ (nint)3);
            z += 4;
        }
```

This would be the way to write that without using pointers directly. Compiles to ( in .net 7 )

```
       movzx    rsi, byte  ptr [rdx+r11]
       add      eax, esi
       movzx    rsi, byte  ptr [rdx+r11+01H]
       add      r8d, esi
       movzx    rsi, byte  ptr [rdx+r11+02H]
       add      r9d, esi
       movzx    rsi, byte  ptr [rdx+r11+03H]
       add      r10d, esi
       add      r11, 4
       cmp      r11d, ecx
       jle      SHORT G_M000_IG06
```